aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -126,4 +126,65 @@ Tensor flip_cuda(const Tensor& self, IntList dims) {
   return out_tensor;
 }
 
+template <typename scalar_t>
+__global__
+void roll_cuda_kernel(scalar_t* in_tensor, scalar_t* out_tensor, int64_t N,
+                      int64_t roll_dim, int64_t shift, int64_t start,
+                      int64_t size, int64_t stride, int64_t total_dims) {
+  int64_t linear_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_index >= N) {
+    return;
+  }
+  // roll dim idx is the index of linear_index along the rolling dimension.
+  int64_t roll_dim_idx = linear_index % (stride * size) / stride;
```

Comment:
something looks broken in here (as an aside -- admittedly don't know the precedence of % compared to /.

But try this:
```
torch.roll(torch.arange(12,device='cuda').view(3,4)[:,0], 0, 0)
```

Reply:
not sure what you mean here, roll by zero doesn't do anything:

>>> x = torch.arange(12, device='cuda').view(3,4)[:,0]
>>> x
tensor([0, 4, 8], device='cuda:0')
>>> x.equal(x.roll(0,0))
True

Reply:
Just to verify with a manual test on a large arbitrary tensor:

>>> x = torch.randn(100,5,5,100,50,4)
>>> x[40, 3, 2, 55, 30, 2]
tensor(-0.0077)
>>> x.stride()
(500000, 100000, 20000, 200, 4, 1)
>>> x.view(x.numel())[40 * 500000 + 3 * 100000 + 2 * 20000 + 55 * 200 + 30 * 4 + 2 * 1 ]
tensor(-0.0077)
>>> index = 40 * 500000 + 3 * 100000 + 2 * 20000 + 55 * 200 + 30 * 4 + 2 * 1
>>> index
20351122

Then in scala (to use C++ mod behavior) using the above logic to convert the linear index back to the individual indices:

scala> val index = 40 * 500000 + 3 * 100000 + 2 * 20000 + 55 * 200 + 30 * 4 + 2 * 1
index: Int = 20351122
scala> val strides = List(500000, 100000, 20000, 200, 4, 1)
strides: List[Int] = List(500000, 100000, 20000, 200, 4, 1)

scala> val sizes = List(100, 5, 5, 100, 50, 4)
sizes: List[Int] = List(100, 5, 5, 100, 50, 4)

scala> (strides zip sizes).map{ case (stride, size) => index % (stride * size) / stride }
res9: List[Int] = List(40, 3, 2, 55, 30, 2)

This logic might be nice to encapsulate and test in some other place, since it might be useful elsewhere, but to do so we'd need to add it as a cuda function. Do we have have build targets to directly test individual internal cuda functions?

Reply:
no, we don't have build targets / bindings to internal cuda functions.